utils/signer/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+use std::str::FromStr;
+
+use alloy_consensus::TxEnvelope;
+use alloy_eips::Decodable2718;
+use alloy_network::{Ethereum, EthereumWallet, TransactionBuilder};
+use alloy_primitives::{Address, Bytes};
+use alloy_provider::{Provider, ProviderBuilder, Web3Signer};
+use alloy_rpc_types_eth::{TransactionReceipt, TransactionRequest};
+use alloy_signer_local::PrivateKeySigner;
+use alloy_transport_http::reqwest::Url;
+use anyhow::{Context, Result};
+use serde::{Deserialize, Serialize};
+use tokio::time::Duration;
+
+pub const NUM_CONFIRMATIONS: u64 = 3;
+pub const TIMEOUT_SECONDS: u64 = 60;
+
+#[derive(Clone, Debug)]
+/// The type of signer to use for signing transactions.
+pub enum Signer {
+    /// The signer URL and address.
+    Web3Signer(Url, Address),
+    /// The local signer.
+    LocalSigner(PrivateKeySigner),
+}
+
+impl Signer {
+    pub fn address(&self) -> Address {
+        match self {
+            Signer::Web3Signer(_, address) => *address,
+            Signer::LocalSigner(signer) => signer.address(),
+        }
+    }
+
+    pub fn from_env() -> Result<Self> {
+        if let (Ok(signer_url_str), Ok(signer_address_str)) =
+            (std::env::var("SIGNER_URL"), std::env::var("SIGNER_ADDRESS"))
+        {
+            let signer_url = Url::parse(&signer_url_str).context("Failed to parse SIGNER_URL")?;
+            let signer_address =
+                Address::from_str(&signer_address_str).context("Failed to parse SIGNER_ADDRESS")?;
+            Ok(Signer::Web3Signer(signer_url, signer_address))
+        } else if let Ok(private_key_str) = std::env::var("PRIVATE_KEY") {
+            let private_key = PrivateKeySigner::from_str(&private_key_str)
+                .context("Failed to parse PRIVATE_KEY")?;
+            Ok(Signer::LocalSigner(private_key))
+        } else {
+            anyhow::bail!(
+                "Neither (SIGNER_URL and SIGNER_ADDRESS) nor PRIVATE_KEY are set in environment"
+            )
+        }
+    }
+}
+
+/// Sign a transaction request using the configured `signer`.
+pub async fn sign_transaction_request_inner(
```

Comment:
This should be in `impl Signer` and take in &self. Also, name/comment are a bit misleading: it says sign but it also sends and awaits the transaction. Maybe it should just be called `send_transaction_request`?